### PR TITLE
Avoid storing layer contents in memory during optimization

### PR DESCRIPTION
**Pull request by ktock**
During optimization, we currently store the layer contents on memory ([]byte)
but it ends up consuming large memory. This commit solves this by storing layers
to temporary files instead.

## Memory usage comparison on `glassfish:4.1-jdk8` optimization

```
ctr-remote image optimize --plain-http glassfish:4.1-jdk8 http://registry2:5000/glassfish:4.1-jdk8
```

### master branch
```
(pprof) top
Showing nodes accounting for 513.71MB, 99.02% of 518.78MB total
Dropped 189 nodes (cum <= 2.59MB)
      flat  flat%   sum%        cum   cum%
  512.02MB 98.70% 98.70%   512.02MB 98.70%  bytes.makeSlice
    1.67MB  0.32% 99.02%     3.32MB  0.64%  github.com/google/crfs/stargz.(*Writer).AppendTar
    0.01MB 0.0023% 99.02%     3.34MB  0.64%  github.com/containerd/stargz-snapshotter/cmd/ctr-remote/commands.(*layer).Compressed.func1
         0     0% 99.02%      512MB 98.69%  bytes.(*Buffer).ReadFrom
         0     0% 99.02%   512.02MB 98.70%  bytes.(*Buffer).grow
         0     0% 99.02%   512.04MB 98.70%  github.com/containerd/stargz-snapshotter/cmd/ctr-remote/commands.optimize.func2
         0     0% 99.02%   512.04MB 98.70%  golang.org/x/sync/errgroup.(*Group).Go.func1
         0     0% 99.02%      512MB 98.69%  io/ioutil.ReadAll
         0     0% 99.02%      512MB 98.69%  io/ioutil.readAll
```

### patched version
```
(pprof) top
Showing nodes accounting for 7229.28kB, 90.42% of 7994.82kB total
Dropped 151 nodes (cum <= 39.97kB)
Showing top 10 nodes out of 95
      flat  flat%   sum%        cum   cum%
 1888.18kB 23.62% 23.62%  4522.74kB 56.57%  github.com/google/crfs/stargz.(*Writer).AppendTar
 1693.47kB 21.18% 44.80%  2433.73kB 30.44%  archive/tar.(*Reader).readHeader
    1296kB 16.21% 61.01%  1440.65kB 18.02%  compress/flate.NewWriter
  740.26kB  9.26% 70.27%   740.26kB  9.26%  archive/tar.(*parser).parseString
  480.66kB  6.01% 76.28%   484.92kB  6.07%  fmt.Sprintf
  350.04kB  4.38% 80.66%   350.04kB  4.38%  io.NewSectionReader
  285.11kB  3.57% 84.23%  2686.67kB 33.61%  github.com/containerd/stargz-snapshotter/cmd/ctr-remote/sorter.importTar
  232.91kB  2.91% 87.14%   232.91kB  2.91%  time.Time.Format
     136kB  1.70% 88.84%      136kB  1.70%  compress/flate.(*compressor).initDeflate
  126.64kB  1.58% 90.42%  1594.09kB 19.94%  io.copyBuffer
```

### CODE REVIEW

File: cmd/ctr-remote/commands/optimize.go [hunk 1]
```diff
@@ -247,33 +248,48 @@ func convert(clicontext *cli.Context, srcRef, dstRef name.Reference, runopts ...
...
 // The order of the "in" list must be base layer first, top layer last.
-func optimize(clicontext *cli.Context, in []regpkg.Layer, config imgpkg.Image, opts ...sampler.Option) (out []regpkg.Layer, err error) {
-	root := ""
+func optimize(clicontext *cli.Context, in []regpkg.Layer, config imgpkg.Image, opts ...sampler.Option) (out []regpkg.Layer, done func(), err error) {
+	// Setup temporary workspace
+	tmpRoot, err := ioutil.TempDir("", "optimize")
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = os.Chmod(tmpRoot, 0777); err != nil {
```
Comment (AkihiroSuda) [inline cmd/ctr-remote/commands/optimize.go:257]: Why?

File: cmd/ctr-remote/commands/optimize.go [hunk 2]
```diff
@@ -247,33 +248,48 @@ func convert(clicontext *cli.Context, srcRef, dstRef name.Reference, runopts ...
...
-		addLayer  = make([](func([]regpkg.Layer) ([]regpkg.Layer, error)), len(in))
+		eg         errgroup.Group
+		lowerdirs  []string
+		addLayer   = make([](func([]regpkg.Layer) ([]regpkg.Layer, error)), len(in))
+		layerFiles []*os.File
 	)
+	done = func() {
+		for _, f := range layerFiles {
+			if err := f.Close(); err != nil {
+				fmt.Printf("Warn: failed to close tmpfile %v: %v\n", f.Name(), err)
```
Comment (AkihiroSuda) [inline cmd/ctr-remote/commands/optimize.go:281]: consider use logrus

File: cmd/ctr-remote/commands/optimize.go (hunk 1 above)
Comment (ktock) [reply cmd/ctr-remote/commands/optimize.go]: Thanks for the comment. Fixed it.

File: cmd/ctr-remote/commands/optimize.go (hunk 2 above)
Comment (ktock) [reply cmd/ctr-remote/commands/optimize.go:278]: I'll open a following-up PR for refactoring around logging of ctr-remote command.

**Pull request merged**